# Agenda

1. Redução de Dimensionalidade
- Seleção de features
- Aplicação Prática

# 1. Redução de Dimensionalidade

<p> Refere-se ao processo de conversão de um conjunto de dados que têm vastas dimensões em dados com dimensões menores</p>

# 2. Seleção de Features:

- Identificar e selecionar features relevantes

## Como:

- Valores Nulos
- Baixa Variância
- Arvores de decisão
- Random Forest
- Alta correlação
- PCA
- Recursive Feature Elimination
- Algoritmo de Minimum Redundancy Maximum Relevance (mRMR);

## Benefícios


 - Isso ajuda a comprimir dados e a reduzir o espaço de armazenamento requerido.
 - Acelera o tempo necessário para a realização de cálculos dos mesmos. Menos dimensões conduz a uma menor computação, menos dimensões também pode permitir o uso de algoritmos que não se aplicariam para o caso de um número de dimensões muito grande.
 - Ela cuida da multi-colinearidade que melhora o desempenho do modelo. Remove recursos redundantes. Por exemplo: não há lógica em armazenar um valor em duas unidades diferentes (em metros e polegadas).
 - Reduzir as dimensões de dados para 2D ou 3D pode nos permitir traçar e visualizar com precisão. Você pode, então, observar padrões mais claramente. Abaixo, você pode ver isso, como um dado 3D é convertida em 2D. Primeiro ele identificou o plano 2D, e então representou os pontos sobre estes dois novos eixos z1 e z2.

# Exemplo Valores Nulos

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

## Base de apoio com os ids das empresas e portfolio referência

In [2]:
companies_ids = pd.read_csv('../../csvs/all_companies_ids.csv')

## Base do mercado incluindo a informação do portfolio referência 

In [3]:
df = pd.read_csv('../../csvs/estaticos_market.csv')

In [4]:
df_merged = df.merge(companies_ids, left_on='id', right_on='id')
df_merged

Unnamed: 0                                                 id  \
0            369  dabe79bec87c88ae04e869bf6bd321ee5e1893cecf6625...   
1            604  32e5f4e10932153a7ba869cb0386e7e02d49d2461046b8...   
2           1646  09e95c1a84049001d086470a8f320a19b076f955a89122...   
3           1703  a95d6f30bba445bd3d6b0c5b36f865b38ec01d17336090...   
4           2521  6cb309685cea0b6d2988818792ec2e6fcb2bd02e0afa9e...   
...          ...                                                ...   
1381      461443  31438f752b4a846638c0acd98c3078d7e1ee83204e0388...   
1382      461483  4fbebb5a4ea9d2c5da87f446aff48403f0409e2e8fe381...   
1383      461834  8a226339b331b5aec24877a69edc40070b9dd7e71b2194...   
1384      461978  8bbff0a749bceb9749c5edd90373270ff9109f069ef27c...   
1385      462190  970485e6b8abe52d8fb4e0bf060fb6f0f4f8576cb7583d...   

      fl_matriz                       de_natureza_juridica sg_uf  \
0          True                      EMPRESARIO INDIVIDUAL    PI   
1         False                      EMPRESARIO INDIVIDUAL    MA   
2          True  AUTARQUIA ESTADUAL OU DO DISTRITO FEDERAL    AC   
3          True                      EMPRESARIO INDIVIDUAL    MA   
4          True                      EMPRESARIO INDIVIDUAL    RN   
...         ...                                        ...   ...   
1381       True                    SERVICO SOCIAL AUTONOMO    RO   
1382       True                      EMPRESARIO INDIVIDUAL    PI   
1383       True              SOCIEDADE EMPRESARIA LIMITADA    PI   
1384       True                      EMPRESARIO INDIVIDUAL    MA   
1385       True                      EMPRESARIO INDIVIDUAL    PI   

            natureza_juridica_macro  \
0                            OUTROS   
1                            OUTROS   
2             ADMINISTRACAO PUBLICA   
3                            OUTROS   
4                            OUTROS   
...                             ...   
1381  ENTIDADES SEM FINS LUCRATIVOS   
1382                         OUTROS   
1383         ENTIDADES EMPRESARIAIS   
1384                         OUTROS   
1385                         OUTROS   

                                                de_ramo             setor  \
0                                    COMERCIO VAREJISTA          COMERCIO   
1                                    COMERCIO VAREJISTA          COMERCIO   
2     ADMINISTRACAO PUBLICA, DEFESA E SEGURIDADE SOCIAL           SERVIÇO   
3                                    COMERCIO VAREJISTA          COMERCIO   
4                                    COMERCIO VAREJISTA          COMERCIO   
...                                                 ...               ...   
1381                               SERVICOS DE EDUCACAO           SERVIÇO   
1382                                 COMERCIO VAREJISTA          COMERCIO   
1383                            INDUSTRIA DA CONSTRUCAO  CONSTRUÇÃO CIVIL   
1384                                 COMERCIO VAREJISTA          COMERCIO   
1385                                 COMERCIO VAREJISTA          COMERCIO   

      idade_empresa_anos idade_emp_cat  ...  max_meses_servicos  \
0               0.646575          <= 1  ...                 NaN   
1               6.586301        5 a 10  ...                 NaN   
2              11.243836       10 a 15  ...          357.066667   
3               8.010959        5 a 10  ...                 NaN   
4              20.863014          > 20  ...                 NaN   
...                  ...           ...  ...                 ...   
1381           18.580822       15 a 20  ...          300.466667   
1382           31.868493          > 20  ...                 NaN   
1383           41.164384          > 20  ...          258.033333   
1384            0.969863          <= 1  ...                 NaN   
1385            3.994521         1 a 5  ...                 NaN   

      min_meses_servicos  qt_funcionarios  qt_funcionarios_12meses  \
0                    NaN              NaN                      NaN 

## % De Dados faltantes por variável

In [5]:
aux = pd.DataFrame({'variables': df_merged.columns,
                    'data_types': df_merged.dtypes,
                    'missing_percentage': df_merged.isna().sum() / df_merged.shape[0]})
aux

variables data_types  missing_percentage
Unnamed: 0                          Unnamed: 0      int64            0.000000
id                                          id     object            0.000000
fl_matriz                            fl_matriz       bool            0.000000
de_natureza_juridica      de_natureza_juridica     object            0.000000
sg_uf                                    sg_uf     object            0.000000
...                                        ...        ...                 ...
tx_crescimento_12meses  tx_crescimento_12meses    float64            0.331169
tx_crescimento_24meses  tx_crescimento_24meses    float64            0.336219
tx_rotatividade                tx_rotatividade    float64            0.305916
qt_filiais                          qt_filiais      int64            0.000000
empresa                                empresa     object            0.000000

[183 rows x 3 columns]

In [6]:
data_missing_zero = aux[aux['missing_percentage']==0]
variables_missing_zero = [variable for variable in data_missing_zero['variables']]

### Das 183 colunas identificamos 32 variaveis com nenhum dado faltante

In [7]:
len(variables_missing_zero)

32

In [8]:
new_data = df_merged[variables_missing_zero]
new_data

Unnamed: 0                                                 id  \
0            369  dabe79bec87c88ae04e869bf6bd321ee5e1893cecf6625...   
1            604  32e5f4e10932153a7ba869cb0386e7e02d49d2461046b8...   
2           1646  09e95c1a84049001d086470a8f320a19b076f955a89122...   
3           1703  a95d6f30bba445bd3d6b0c5b36f865b38ec01d17336090...   
4           2521  6cb309685cea0b6d2988818792ec2e6fcb2bd02e0afa9e...   
...          ...                                                ...   
1381      461443  31438f752b4a846638c0acd98c3078d7e1ee83204e0388...   
1382      461483  4fbebb5a4ea9d2c5da87f446aff48403f0409e2e8fe381...   
1383      461834  8a226339b331b5aec24877a69edc40070b9dd7e71b2194...   
1384      461978  8bbff0a749bceb9749c5edd90373270ff9109f069ef27c...   
1385      462190  970485e6b8abe52d8fb4e0bf060fb6f0f4f8576cb7583d...   

      fl_matriz                       de_natureza_juridica sg_uf  \
0          True                      EMPRESARIO INDIVIDUAL    PI   
1         False                      EMPRESARIO INDIVIDUAL    MA   
2          True  AUTARQUIA ESTADUAL OU DO DISTRITO FEDERAL    AC   
3          True                      EMPRESARIO INDIVIDUAL    MA   
4          True                      EMPRESARIO INDIVIDUAL    RN   
...         ...                                        ...   ...   
1381       True                    SERVICO SOCIAL AUTONOMO    RO   
1382       True                      EMPRESARIO INDIVIDUAL    PI   
1383       True              SOCIEDADE EMPRESARIA LIMITADA    PI   
1384       True                      EMPRESARIO INDIVIDUAL    MA   
1385       True                      EMPRESARIO INDIVIDUAL    PI   

            natureza_juridica_macro  \
0                            OUTROS   
1                            OUTROS   
2             ADMINISTRACAO PUBLICA   
3                            OUTROS   
4                            OUTROS   
...                             ...   
1381  ENTIDADES SEM FINS LUCRATIVOS   
1382                         OUTROS   
1383         ENTIDADES EMPRESARIAIS   
1384                         OUTROS   
1385                         OUTROS   

                                                de_ramo             setor  \
0                                    COMERCIO VAREJISTA          COMERCIO   
1                                    COMERCIO VAREJISTA          COMERCIO   
2     ADMINISTRACAO PUBLICA, DEFESA E SEGURIDADE SOCIAL           SERVIÇO   
3                                    COMERCIO VAREJISTA          COMERCIO   
4                                    COMERCIO VAREJISTA          COMERCIO   
...                                                 ...               ...   
1381                               SERVICOS DE EDUCACAO           SERVIÇO   
1382                                 COMERCIO VAREJISTA          COMERCIO   
1383                            INDUSTRIA DA CONSTRUCAO  CONSTRUÇÃO CIVIL   
1384                                 COMERCIO VAREJISTA          COMERCIO   
1385                                 COMERCIO VAREJISTA          COMERCIO   

      idade_empresa_anos idade_emp_cat  ...  fl_spa  fl_antt  fl_veiculo  \
0               0.646575          <= 1  ...   False    False       False   
1               6.586301        5 a 10  ...   False    False       False   
2              11.243836       10 a 15  ...   False    False        True   
3               8.010959        5 a 10  ...   False    False       False   
4              20.863014          > 20  ...   False    False       False   
...                  ...           ...  ...     ...      ...         ...   
1381           18.580822       15 a 20  ...   False    False        True   
1382           31.868493          > 20  ...   False    False       False   
1383           41.164384          > 20  ...   False    False        True   
1384            0.969863          <= 1  ...   False    False       False   
1385            3.994521         1 a 5  ...   False    False       False   

      vl_total_veiculos_pesad

# Variaveis por tipo em ordem decrescente em % de dados faltantes

In [9]:
pd.set_option("max_rows", None)
aux = pd.DataFrame({'variables': new_data.columns,
                    'var_types': new_data.dtypes,
                    'missing_percentages': new_data.isna().sum() / new_data.shape[0]})
pivot = pd.pivot_table(aux, index=['var_types', 'variables'])
pivot.sort_values(by=['var_types','missing_percentages'],  ascending=False)

missing_percentages
var_types variables                                           
object    de_natureza_juridica                             0.0
          de_ramo                                          0.0
          dt_situacao                                      0.0
          empresa                                          0.0
          fl_antt                                          0.0
          fl_passivel_iss                                  0.0
          fl_rm                                            0.0
          fl_simples_irregular                             0.0
          fl_spa                                           0.0
          fl_veiculo                                       0.0
          id                                               0.0
          idade_emp_cat                                    0.0
          natureza_juridica_macro                          0.0
          nm_divisao                                       0.0
          nm_segmento                                      0.0
          setor                                            0.0
          sg_uf                                            0.0
          sg_uf_matriz                                     0.0
float64   idade_empresa_anos                               0.0
          vl_total_veiculos_leves_grupo                    0.0
          vl_total_veiculos_pesados_grupo                  0.0
int64     Unnamed: 0                                       0.0
          qt_filiais                                       0.0
bool      fl_email                                         0.0
          fl_epp                                           0.0
          fl_ltda                                          0.0
          fl_matriz                                        0.0
          fl_me                                            0.0
          fl_mei                                           0.0
          fl_sa                                            0.0
          fl_st_especial                                   0.0
          fl_telefone                                      0.0

## Referências

* [Guia para redução de dimensionalidade](https://www.vooo.pro/insights/guia-de-aprendizado-para-iniciantes-em-tecnicas-de-reducao-de-dimensoes-dimension-reduction/)